Src/FluentAssertions/Specialized/TaskCompletionSourceAssertions.cs

```diff
@@ -8,6 +8,105 @@ namespace FluentAssertions.Specialized;
 
 #pragma warning disable CS0659 // Ignore not overriding Object.GetHashCode()
 #pragma warning disable CA1065 // Ignore throwing NotSupportedException from Equals
+
+#if NET6_0_OR_GREATER
+
+public class TaskCompletionSourceAssertions
+{
+    private readonly TaskCompletionSource subject;
+
+    public TaskCompletionSourceAssertions(TaskCompletionSource tcs)
+        : this(tcs, new Clock())
+    {
+    }
+
+    public TaskCompletionSourceAssertions(TaskCompletionSource tcs, IClock clock)
+    {
+        subject = tcs;
+        Clock = clock ?? throw new ArgumentNullException(nameof(clock));
+    }
+
+    private protected IClock Clock { get; }
+
+    /// <summary>
+    /// Asserts that the <see cref="Task"/> of the current <see cref="TaskCompletionSource{T}"/> will complete within the specified time.
+    /// </summary>
+    /// <param name="timeSpan">The allowed time span for the operation.</param>
+    /// <param name="because">
+    /// A formatted phrase as is supported by <see cref="string.Format(string,object[])" /> explaining why the assertion
+    /// is needed. If the phrase does not start with the word <i>because</i>, it is prepended automatically.
+    /// </param>
+    /// <param name="becauseArgs">
+    /// Zero or more objects to format using the placeholders in <paramref name="because" />.
+    /// </param>
+    public async Task CompleteWithinAsync(
+        TimeSpan timeSpan, string because = "", params object[] becauseArgs)
+    {
+        Execute.Assertion
+            .ForCondition(subject is not null)
```

Comment:
I have thought about a possible solution, but can't find one.
What kind of `AndWhichConstraint` can I return for useful addition assertions?

Reply:
I don't think there's any useful `AndWhichConstraint` to return here, just `AndConstraint`.

A test to expose this could be something like this
(I'm rewriting `When_null_be_writable_should_fail` directly on GH, so hope it's valid code :crossed_fingers: )

```cs
// Arrange
TaskCompletionSource tcs = null;

// Act
Func<Task> act = async () =>
{
    using var _ = new AssertionScope();
    await tcs.Should().CompleteWithinAsync("we want to test the failure {0}", "message");
};

// Assert
(await act.Should().ThrowAsync<XunitException>())
    .WithMessage("Expected tcs to be complete within *failure message*, but found <null>.");
```

Reply:
> A test to expose this could be something like this
> (I'm rewriting `When_null_be_writable_should_fail` directly on GH, so hope it's valid code 🤞 )

My question was unclear: I know how to test it, but I do not how to implement this. I have not understand today how `AssertionScope` is working and what I have to take care about.

Reply:
Roger.

When writing 
```cs
using (new AssertionScope)
{
    true.Should().Be(false);
    "foo".Should().Be("bar");
}
```

The test will not fail at first assertion, but continue until the `AssertionScope` is disposed, which then checks if any assertions failed and _then_ fail the test.

So the pitfall is
```cs
TaskCompletionSource tcs = null;

using (new AssertionScope)
{
    await tcs.Should().CompleteWithinAsync(default);
}
```

The assertion inside `CompleteWithinAsync` checking `subject is not null` will add an assertion failure to the `AssertionScope`, but execution continues and we throw an NRE when a little later invoking `subject.Task`.

Here's an example of how to guard against this.
https://github.com/fluentassertions/fluentassertions/blob/cb5a34a29caedc273e9d58c6d814bff2c13a13c5/Src/FluentAssertions/Streams/StreamAssertions.cs#L45-L61

Reply:
Once again partial misunderstanding: I know how to **use** `AssertionScope` but I did not know how to take care **internally**.
Your hint to `StreamAssertions.cs` is helpful, I guess...